### PR TITLE
Moved ThreadCount initialization up the pipeline

### LINKED ISSUE #404: The default thread count should be reflected on KestrelServerInformation.ThreadCount during application start
The default thread count is currently determined by `KestrelEngine` **after** the application's `Conifgure` method has been called. This means the application can't inspect the default value for `IKestrelServerInformation.ThreadCount`, like it can for `IKestrelServerInformation.NoDelay`. Essentially, the following code should echo the correct defaults:

``` C#
public void Configure(IApplicationBuilder app)
{
    var kestrel = app.ServerFeatures[typeof(IKestrelServerInformation)] as IKestrelServerInformation;

    if (kestrel != null)
    {
        Console.WriteLine($"Kestrel thread count: {kestrel.ThreadCount}");
        Console.WriteLine($"Kestrel TCP no-delay: {kestrel.NoDelay}");
    }
```

### PR DESCRIPTION
**Pull request by khellang**
Closes #404. Replaces #397.

### CODE REVIEW

**Comment by muratg** [comment]
@halter73 Could you have a look?

File: src/Microsoft.AspNet.Server.Kestrel/KestrelServer.cs [hunk 1]
```diff
@@ -75,35 +75,14 @@ public void Start(RequestDelegate requestDelegate)
...
 
-                // Actual core count would be a better number 
-                // rather than logical cores which includes hyper-threaded cores.
-                // Divide by 2 for hyper-threading, and good defaults (still need threads to do webserving).
-                // Can be user overriden using IKestrelServerInformation.ThreadCount
-                var threadCount = Environment.ProcessorCount >> 1;
+                var threadCount = information.ThreadCount;
 
-                if (threadCount < 1)
+                if (threadCount < 0)
```
Comment (halter73) [inline src/Microsoft.AspNet.Server.Kestrel/KestrelServer.cs]: This PR looks good, sorry for the delayed response. The only thing I would change is making this `threadCount <= 0` and the exception message ""ThreadCount must be positive."

Comment (khellang) [inline src/Microsoft.AspNet.Server.Kestrel/KestrelServer.cs]: Fixed! :+1:

**Comment by khellang** [comment]
Squashed and ready to go :wink:

**Comment by halter73** [comment]
Thanks!

**Pull request merged**